tests/mapper/test_link_mappers.py

```diff
@@ -783,3 +783,390 @@ def test_Attri2VecLinkGenerator_unsupervisedSampler_sample_generation(self):
+
+            for ii in range(6, 8):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.in_samples[0] * self.in_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(8, 10):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.in_samples[0] * self.out_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(10, 12):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.out_samples[0] * self.in_samples[1],
+                    self.n_feat,
+                )
+
+            for ii in range(12, 14):
+                assert nf[ii].shape == (
+                    min(self.batch_size, data_size),
+                    self.out_samples[0] * self.out_samples[1],
+                    self.n_feat,
+                )
+
+        with pytest.raises(IndexError):
+            nf, nl = mapper[2]
+
+    def test_DirectedGraphSAGELinkGenerator_shuffle(self):
+        def test_edge_consistency(shuffle):
+            G = example_graph(feature_size=1, is_directed=True)
+            edges = list(G.edges())
+            edge_labels = list(range(len(edges)))
+
+            mapper = DirectedGraphSAGELinkGenerator(
+                G, batch_size=2, in_samples=[0], out_samples=[0]
+            ).flow(edges, edge_labels, shuffle=shuffle)
+
+            assert len(mapper) == 2
+
+            for batch in range(len(mapper)):
+                nf, nl = mapper[batch]
+                e1 = edges[nl[0]]
+                e2 = edges[nl[1]]
+                assert nf[0][0, 0, 0] == e1[0]
+                assert nf[1][0, 0, 0] == e1[1]
+                assert nf[0][1, 0, 0] == e2[0]
+                assert nf[1][1, 0, 0] == e2[1]
+
+        test_edge_consistency(True)
+        test_edge_consistency(False)
+
+    def test_DirectedGraphSAGELinkGenerator_not_Stellargraph(self):
+        G = nx.Graph()
+        elist = [(1, 2), (2, 3), (1, 4), (3, 2)]
+        G.add_edges_from(elist)
+
+        # Add example features
+        for v in G.nodes():
+            G.nodes[v]["feature"] = np.ones(1)
+
+        with pytest.raises(TypeError):
+            DirectedGraphSAGELinkGenerator(
+                G,
+                batch_size=self.batch_size,
+                in_samples=self.in_samples,
+                out_samples=self.out_samples,
+            )
+
+    def test_DirectedGraphSAGELinkGenerator_zero_samples(self):
+
+        G = example_graph(feature_size=self.n_feat, is_directed=True)
+        data_size = G.number_of_edges()
+        edge_labels = [0] * data_size
+
+        mapper = DirectedGraphSAGELinkGenerator(
+            G, batch_size=self.batch_size, in_samples=[0], out_samples=[0],
+        ).flow(G.edges(), edge_labels)
+
+        assert len(mapper) == 2
+
+        for ii in range(1):
```

Comment:
Maybe you meant to do:
```suggestion
        for ii in range(len(mapper)):
```
unless you really wanted to only check the first element, which means we don't need the for loop at all
```python
nf, nl = mapper[0]
...
```

Reply:
uh.. thats meant to loop through them all!